From cce320e1af951a246b2ccc72714dc5478b3907ca Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 8 Aug 2018 12:56:50 +0300
Subject: [PATCH 0541/1051] net: mvpp2: fix 10G-SFP link status interrupt
 allocation

PROBLEM:
On 8040-DB/7040-DB boards the 10G SFP managed "link" status interrupt
is present but doesn't work. So 10G devices are not Up, or randomly Up
and do not detect the fiber-cable connect/disconnect.

Root-cause:
Wrong condition deciding to Not allocate link-status interrupt.

Solution:
Go to allocate the interrupt if phylink is present but the "has_phy"
is FALSE. This is exactly the case of the SFP.

Change-Id: Ie1dacafd20a1ef48f0b8ccb32ecad2a56e236bad
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/59471
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index ff1250cae3a4..b0f2b85d8291 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3510,7 +3510,8 @@ static int mvpp2_open(struct net_device *dev)
 		valid = true;
 	}
 
-	if (priv->hw_version != MVPP21 && port->link_irq && !port->phylink) {
+	if (priv->hw_version != MVPP21 && port->link_irq &&
+	    (!port->phylink || !port->has_phy)) {
 		mvpp2_txqs_on_tasklet_init(port);
 		err = request_irq(port->link_irq, mvpp2_link_status_isr, 0,
 				  dev->name, port);
-- 
2.17.1

